src/v/cloud_storage_clients/client_pool.h

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2020 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "cloud_roles/apply_credentials.h"
+#include "cloud_storage_clients/client.h"
+#include "utils/gate_guard.h"
+#include "utils/intrusive_list_helpers.h"
+
+#include <seastar/core/condition-variable.hh>
+
+namespace cloud_storage_clients {
+
+/// Policy that controls behaviour of the client pool
+/// in situation when number of requested client connections
+/// exceeds pool capacity
+enum class client_pool_overdraft_policy {
```

Comment:
It might be a good time to think about removing this policy. We always use `wait_if_empty` behavior.